File: typescript/src/tokenizer.ts
/**
 * A token with metadata about how it was originally written.
 */
export interface TokenMeta {
  text: string;
  wasQuoted: boolean;
}

/**
 * Tokenize an operation string by whitespace, respecting quoted strings.
 * Returns structured tokens that preserve whether each token was originally quoted.
 *
 * This allows downstream code (e.g. parseOp) to skip key:value classification
 * for quoted tokens like "LTV:CAC".
 */
export function tokenizeWithMeta(input: string): TokenMeta[] {
  const tokens: TokenMeta[] = [];
  let i = 0;
  const len = input.length;

  while (i < len) {
    // Skip whitespace
    while (i < len && input[i] === " ") i++;
    if (i >= len) break;

    if (input[i] === '"') {
      // Quoted string
      i++; // skip opening quote
      let token = "";
      while (i < len && input[i] !== '"') {
        if (input[i] === "\\" && i + 1 < len) {
          const next = input[i + 1];
          if (next === "n") {
            token += "\n";
            i += 2;
          } else if (next === "u") {
            // \uXXXX — read 4 hex digits
            const hex = input.slice(i + 2, i + 6);
            if (hex.length === 4 && /^[0-9a-fA-F]{4}$/.test(hex)) {
              token += String.fromCharCode(parseInt(hex, 16));
              i += 6;
            } else {
              // Invalid hex — pass through literally
              token += "\\u";
              i += 2;
            }
          } else {
            i++;
            token += input[i];
            i++;
          }
        } else {
          token += input[i];
          i++;
        }
      }
      if (i < len) i++; // skip closing quote
      tokens.push({ text: token, wasQuoted: true });
    } else {
      // Unquoted token — preserve embedded quotes (e.g., key:"value" → key:"value")
      let token = "";
      while (i < len && input[i] !== " ") {
        if (input[i] === '"') {
          // Embedded quoted value — preserve quotes in token for downstream detection
          token += '"';
          i++; // skip opening quote
          while (i < len && input[i] !== '"') {
            if (input[i] === "\\" && i + 1 < len) {
              const next = input[i + 1];
              if (next === "n") {
                token += "\n";
                i += 2;
              } else if (next === "u") {
                // \uXXXX — read 4 hex digits
                const hex = input.slice(i + 2, i + 6);
                if (hex.length === 4 && /^[0-9a-fA-F]{4}$/.test(hex)) {
                  token += String.fromCharCode(parseInt(hex, 16));
                  i += 6;
                } else {
                  // Invalid hex — pass through literally
                  token += "\\u";
                  i += 2;
                }
              } else {
                i++;
                token += input[i];
                i++;
              }
            } else {
              token += input[i];
              i++;
            }
          }
          if (i < len) {
            token += '"';
            i++; // skip closing quote
          }
        } else {
          token += input[i];
          i++;
        }
      }
      // Convert literal \n and \uXXXX in unquoted tokens
      tokens.push({ text: token.replace(/\\n/g, "\n").replace(/\\u([0-9a-fA-F]{4})/g, (_, hex) => String.fromCharCode(parseInt(hex, 16))), wasQuoted: false });
    }
  }

  return tokens;
}

/**
 * Tokenize an operation string by whitespace, respecting quoted strings.
 * "add svc \"Auth Service\" theme:blue" -> ["add", "svc", "Auth Service", "theme:blue"]
 */
export function tokenize(input: string): string[] {
  return tokenizeWithMeta(input).map(t => t.text);
}

/**
 * Regex patterns for cell range detection (spreadsheet A1 notation).
 * Cell ref: 1-3 letters followed by digits (A1, BB23, XFD1048576)
 * Row ref: digits only (1, 23)
 *
 * Note: Pure column ranges (A:E, B:B) are intentionally NOT detected here
 * because they are ambiguous with key:value pairs like "theme:blue" or
 * "vel:mf". Column-only ranges should use hyphen syntax (A-E) or be
 * handled at the domain level.
 */
const CELL_REF_RE = /^[A-Za-z]{1,3}\d+$/;
const ROW_REF_RE = /^[0-9]+$/;

/**
 * Check if a token looks like a spreadsheet cell range.
 *
 * Recognized patterns (with optional Sheet! prefix):
 *   A1:F1     — cell range (letters+digits : letters+digits)
 *   3:3       — row range (digits : digits)
 *   1:5       — row range
 *   Sheet2!A1:B10 — cross-sheet cell range
 *
 * NOT recognized (ambiguous with key:value):
 *   A:E       — column range (use A-E instead, or handle at domain level)
 */
export function isCellRange(token: string): boolean {
  let ref = token;
  // Strip optional sheet prefix (Sheet2!A1:B10 → A1:B10)
  const bangIdx = ref.indexOf("!");
  if (bangIdx >= 0) {
    ref = ref.slice(bangIdx + 1);
  }

  const colonIdx = ref.indexOf(":");
  if (colonIdx <= 0 || colonIdx >= ref.length - 1) return false;

  const left = ref.slice(0, colonIdx);
  const right = ref.slice(colonIdx + 1);

  // Cell range: A1:F1 (most common spreadsheet range pattern)
  if (CELL_REF_RE.test(left) && CELL_REF_RE.test(right)) return true;
  // Row range: 1:5 or 3:3 (no FCP key is ever a pure number)
  if (ROW_REF_RE.test(left) && ROW_REF_RE.test(right)) return true;

  return false;
}

/**
 * Check if a token is a key:value pair.
 * Must contain ":" but not start with "@" (selectors), not be an arrow,
 * not be a formula (starts with "="), and not be a cell range (e.g. A1:F1).
 */
export function isKeyValue(token: string): boolean {
  if (token.startsWith("@")) return false;
  if (isArrow(token)) return false;
  // Formulas (=SUM(A1:B2)) are values, not key:value pairs
  if (token.startsWith("=")) return false;
  // Spreadsheet cell ranges (A1:F1, B:B, 3:3) are positional args
  if (isCellRange(token)) return false;
  const colonIdx = token.indexOf(":");
  return colonIdx > 0 && colonIdx < token.length - 1;
}

/**
 * Parse a key:value token. The value may include colons (e.g., "style:orthogonal").
 * Strips surrounding quotes from the value for backwards compatibility.
 */
export function parseKeyValue(token: string): { key: string; value: string } {
  const colonIdx = token.indexOf(":");
  let value = token.slice(colonIdx + 1);
  // Strip surrounding quotes preserved by tokenizer
  if (value.startsWith('"') && value.endsWith('"') && value.length >= 2) {
    value = value.slice(1, -1);
  }
  return {
    key: token.slice(0, colonIdx),
    value,
  };
}

/**
 * Parse a key:value token with metadata about quoting.
 * Returns the unquoted value plus a `wasQuoted` flag.
 */
export function parseKeyValueWithMeta(token: string): { key: string; value: string; wasQuoted: boolean } {
  const colonIdx = token.indexOf(":");
  let value = token.slice(colonIdx + 1);
  let wasQuoted = false;
  if (value.startsWith('"') && value.endsWith('"') && value.length >= 2) {
    value = value.slice(1, -1);
    wasQuoted = true;
  }
  return {
    key: token.slice(0, colonIdx),
    value,
    wasQuoted,
  };
}

/**
 * Check if a token is an arrow operator.
 */
export function isArrow(token: string): boolean {
  return token === "->" || token === "<->" || token === "--";
}

/**
 * Check if a token is a selector (@-prefixed).
 */
export function isSelector(token: string): boolean {
  return token.startsWith("@");
}

File: python/src/fcp_core/tokenizer.py
"""Quote-aware tokenizer for FCP op strings.

Splits on whitespace but respects quoted strings (single and double quotes).
Handles embedded quotes in key:value tokens (e.g. ``title:"Score Chart"``).
Provides helpers for key:value token detection and parsing.
"""

from __future__ import annotations

import re
from dataclasses import dataclass

_WHITESPACE = frozenset(" \t\n\r")


@dataclass
class TokenMeta:
    """A token with metadata about how it was originally written."""

    text: str
    was_quoted: bool


def _consume_quoted(s: str, i: int, quote_char: str) -> tuple[str, int]:
    """Consume characters until the matching *quote_char*, starting after it.

    Handles backslash escape sequences:
      ``\\"`` → literal ``"``   (escaped quote)
      ``\\n`` → newline
      ``\\\\`` → literal ``\\``
      ``\\x`` → literal ``x``  (any other char after backslash)

    Returns ``(content, new_index)`` where *content* excludes the delimiters.
    Raises :class:`ValueError` on unclosed quote.
    """
    n = len(s)
    i += 1  # skip opening quote
    buf: list[str] = []
    while i < n:
        ch = s[i]
        if ch == "\\" and i + 1 < n:
            nxt = s[i + 1]
            if nxt == "n":
                buf.append("\n")
                i += 2
            elif nxt == "u":
                # \uXXXX — read 4 hex digits
                hex_str = s[i + 2 : i + 6]
                if len(hex_str) == 4 and all(c in "0123456789abcdefABCDEF" for c in hex_str):
                    buf.append(chr(int(hex_str, 16)))
                    i += 6
                else:
                    # Invalid hex — pass through literally
                    buf.append("\\u")
                    i += 2
            else:
                # \" → " , \\ → \ , \x → x
                buf.append(nxt)
                i += 2
        elif ch == quote_char:
            return "".join(buf), i + 1  # skip closing quote
        else:
            buf.append(ch)
            i += 1
    raise ValueError("No closing quotation")


def tokenize_with_meta(op_string: str) -> list[TokenMeta]:
    """Split *op_string* on whitespace, respecting quoted substrings.

    Handles three quoting scenarios:

    1. **Standalone quotes** — token starts with ``"`` or ``'``.
       Quotes are stripped, ``was_quoted=True``.
       ``"LTV:CAC"`` → ``TokenMeta(text='LTV:CAC', was_quoted=True)``

    2. **Embedded quotes** — quote appears mid-token (e.g. after ``:``)
       in a ``key:"value with spaces"`` pattern.  The quotes are preserved
       in the token text so ``parse_key_value`` can strip them later.
       ``was_quoted=False``.
       ``title:"Score Chart"`` → ``TokenMeta(text='title:"Score Chart"', was_quoted=False)``

    3. **No quotes** — plain token, ``was_quoted=False``.

    Raises :class:`ValueError` on unclosed standalone quotes.
    """
    tokens: list[TokenMeta] = []
    i = 0
    n = len(op_string)

    while i < n:
        # Skip whitespace
        while i < n and op_string[i] in _WHITESPACE:
            i += 1
        if i >= n:
            break

        ch = op_string[i]

        if ch in ('"', "'"):
            # Standalone quoted string
            content, i = _consume_quoted(op_string, i, ch)
            tokens.append(TokenMeta(text=content, was_quoted=True))
        else:
            # Unquoted token — accumulate until whitespace, handling
            # embedded quotes (e.g. title:"Score Chart") as sub-regions
            buf: list[str] = []
            while i < n and op_string[i] not in _WHITESPACE:
                ch = op_string[i]
                if ch in ('"', "'"):
                    # Embedded quote — preserve delimiters in token text
                    buf.append(ch)
                    i += 1
                    while i < n and op_string[i] != ch:
                        if op_string[i] == "\\" and i + 1 < n:
                            nxt = op_string[i + 1]
                            if nxt == "n":
                                buf.append("\n")
                                i += 2
                            elif nxt == "u":
                                # \uXXXX — read 4 hex digits
                                hex_str = op_string[i + 2 : i + 6]
                                if len(hex_str) == 4 and all(
                                    c in "0123456789abcdefABCDEF" for c in hex_str
                                ):
                                    buf.append(chr(int(hex_str, 16)))
                                    i += 6
                                else:
                                    # Invalid hex — pass through literally
                                    buf.append("\\u")
                                    i += 2
                            else:
                                buf.append(nxt)
                                i += 2
                        else:
                            buf.append(op_string[i])
                            i += 1
                    if i < n:
                        buf.append(op_string[i])  # closing quote
                        i += 1
                else:
                    buf.append(ch)
                    i += 1
            # Convert literal \n and \uXXXX in unquoted tokens
            text = "".join(buf).replace("\\n", "\n")
            text = re.sub(r"\\u([0-9a-fA-F]{4})", lambda m: chr(int(m.group(1), 16)), text)
            tokens.append(TokenMeta(text=text, was_quoted=False))

    return tokens


def tokenize(op_string: str) -> list[str]:
    """Split *op_string* on whitespace, respecting quoted substrings.

    Examples
    --------
    >>> tokenize('add svc "My Service" theme:blue')
    ['add', 'svc', 'My Service', 'theme:blue']
    >>> tokenize("add svc 'My Service' theme:blue")
    ['add', 'svc', 'My Service', 'theme:blue']
    """
    return [t.text for t in tokenize_with_meta(op_string)]


# Patterns for cell range detection (spreadsheet A1 notation).
# Cell ref: 1-3 letters followed by digits (e.g. A1, BB23, XFD1048576)
_CELL_REF_RE = re.compile(r"^[A-Za-z]{1,3}\d+$")
# Row ref: digits only (e.g. 1, 23)
_ROW_REF_RE = re.compile(r"^[0-9]+$")

# Note: Pure column ranges (A:E, B:B) are intentionally NOT detected here
# because they are ambiguous with key:value pairs like "theme:blue" or
# "vel:mf". Column-only ranges should use hyphen syntax (A-E) or be
# handled at the domain level.


def _is_cell_range(token: str) -> bool:
    """Return True if *token* looks like a spreadsheet cell range.

    Recognized patterns (with optional ``Sheet!`` prefix):
      A1:F1     — cell range (letters+digits : letters+digits)
      3:3       — row range (digits : digits)
      1:5       — row range
      Sheet2!A1:B10 — cross-sheet cell range

    NOT recognized (ambiguous with key:value):
      A:E       — column range (use A-E instead, or handle at domain level)
    """
    ref = token
    # Strip optional sheet prefix (Sheet2!A1:B10 → A1:B10)
    if "!" in ref:
        ref = ref.split("!", 1)[1]

    if ":" not in ref:
        return False

    left, right = ref.split(":", 1)
    if not left or not right:
        return False

    # Cell range: A1:F1 (most common spreadsheet range pattern)
    if _CELL_REF_RE.match(left) and _CELL_REF_RE.match(right):
        return True
    # Row range: 1:5 or 3:3 (no FCP key is ever a pure number)
    if _ROW_REF_RE.match(left) and _ROW_REF_RE.match(right):
        return True

    return False


def is_key_value(token: str) -> bool:
    """Return True if *token* is a ``key:value`` pair.

    A key:value token contains ``:``, does NOT start with ``@``
    (that is a selector), is not an arrow (``->``), is not a
    formula (starts with ``=``), and is not a spreadsheet cell
    range (e.g. ``A1:F1``, ``B:B``, ``3:3``).
    """
    if token.startswith("@"):
        return False
    if "->" in token:
        return False
    # Formulas (=SUM(A1:B2)) are values, not key:value pairs
    if token.startswith("="):
        return False
    # Spreadsheet cell ranges (A1:F1, B:B, 3:3) are positional args
    if _is_cell_range(token):
        return False
    return ":" in token


def parse_key_value(token: str) -> tuple[str, str]:
    """Split *token* on the first ``:`` and return ``(key, value)``."""
    key, _, value = token.partition(":")
    # Strip surrounding quotes preserved by posix=False tokenization
    if len(value) >= 2 and value[0] == value[-1] and value[0] in ('"', "'"):
        value = value[1:-1]
    return key, value


def is_selector(token: str) -> bool:
    """Return True if *token* is a selector (starts with ``@``)."""
    return token.startswith("@")


def is_arrow(token: str) -> bool:
    """Return True if *token* is an arrow (``->``, ``<->``, or ``--``)."""
    return token in ("->", "<->", "--")
